### PR TITLE
devicemapper: rework logging and add --storage-opt dm.libdm_log_level

### PR DESCRIPTION
**Pull request by cyphar**
Because libdm doesn't log anything if we have a logging callback
registered, we have to forward our logs to logrus. While this was done
previously in e2f8fbfbc ("devicemapper: split out devicemapper bindings"),
it was modified in e07d3cd9a ("devmapper: Fix libdm logging") to remove
the debugging information because it was too voluminous.

However, this has caused us several issues in debugging production
systems because libdm _LOG_{NOTICE,INFO,DEBUG} logs contain information
that cannot be obtained without recompiling the source.

In order to aid debugging while avoiding large amounts of logs, forward
_LOG_{NOTICE,INFO} into our debug logs and only output _LOG_DEBUG if an
additional storage-driver specific flag is specified. This respects the
fact that most users aren't going to be debugging devicemapper issues,
while also avoiding the source recompilation burden when debugging.

[kitteh](https://flic.kr/p/7fWQeE)

###### [Cosy cat](https://flic.kr/p/7fWQeE) by [David O&#x27;Hare](https://www.flickr.com/photos/30884177@N08/), on Flickr.

Signed-off-by: Aleksa Sarai <asarai@suse.de>

/cc @vbatts @rhvgoyal

### CODE REVIEW

**Comment by cyphar** [comment]
As an aside, I wanted to update the docs for `dockerd` to include the new storagedriver option. I was trying to figure out where the man pages live and it looks like they're no longer in this repo? Where are they now?

**Comment by cpuguy83** [comment]
This looks pretty much like a dup of #32541

**Comment by cyphar** [comment]
I didn't see that change beforehand.

However, this change actually does a lot more than #32541. In particular:

* It cleans up the exposed callback code (which didn't make sense because we always needed to register a callback anyway).
* Cleans up the C callback code to not be limited to 256 bytes.
* Allows a user to specify what their choice of maximum logging is (which is the main point of this patch).
* Removes some of the other old cruft from the logging code that #32541 didn't touch.

I do like some of the stuff in `p/d/log.go` though, we might want to merge those changes as well.

**Comment by cpuguy83** [comment]
Maybe we can just pull those changes in here? (minus the logrus dep in the `p/d` package).

**Comment by cyphar** [comment]
Sure, will do. I'm splitting up this patch now so it's more obvious what I'm doing (I squashed it pre-emptively because I was planning on backporting it).

**Comment by cyphar** [comment]
Actually I just realised that most of the code in `pkg/devicemapper/log.go` is not really useful if you don't want to have a `logrus` dependency inside that file (given the rest of my patchset).

**Comment by rhvgoyal** [comment]
I have not looked at this yet. But we very much require something like this as well. Recompiling source code does not work when debugging a system in production. Restarting docker with additional flag is much easier.

So yes, we do need a knob to get fine control of logging by libdevmapper when used with docker.

File: pkg/devicemapper/devmapper_log.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper_log.go]: So this changes default from fatal to level Err? When I had tested last time, even with this level libdm was spewing tons of messages. And it was made worse because we try device removal in a loop. Most of the time these messages are really not worth it. 

How about keeping the current defaults and providing a knob to change that default.

File: daemon/graphdriver/devmapper/deviceset.go
Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/deviceset.go]: Will it make sense to call it `dm.libdm_log_level` instead.

**Comment by rhvgoyal** [comment]
BTW, I love the idea of breaking down patches in a series. That way one can focus on core bits first and then focus on less important bits later. And its much easier to review.

**Comment by thaJeztah** [comment]
> As an aside, I wanted to update the docs for dockerd to include the new storagedriver option. I was trying to figure out where the man pages live and it looks like they're no longer in this repo? Where are they now?

The CLI docs were moved to the https://github.com/docker/cli repository (yes, `dockerd` _could_ have been left in this repository, but after debating the pros/cons we decided that it made things easier  to have them in that repository 😅)

File: daemon/graphdriver/devmapper/deviceset.go
Comment (cyphar) [inline daemon/graphdriver/devmapper/deviceset.go]: Yeah, that sounds better.

File: pkg/devicemapper/devmapper_log.go
Comment (cyphar) [inline pkg/devicemapper/devmapper_log.go]: I ran this a few times and it didn't seem to spam the logs too much (and I don't like ignoring errors because it causes you to ignore some `_do_dm_ioctl` errors) but yeah, I'll change it back to the old default.

**Comment by cyphar** [comment]
Alright, I'll match a PR there once this is merged.

**Comment by cyphar** [comment]
@rhvgoyal Fix'd. PTAL.

/cc @vbatts

File: daemon/graphdriver/devmapper/deviceset.go
Comment (vrothberg) [inline daemon/graphdriver/devmapper/deviceset.go]: Maybe, we should return a more expressive error here, as for instance `case "dm.thinp_autoextend_threshold":` does.

File: daemon/graphdriver/devmapper/deviceset.go
Comment (vrothberg) [inline daemon/graphdriver/devmapper/deviceset.go]: Considering that `DmLogLevel(value int)` does bounds checking, we could just parse a 64bit end here and defer bounds errors to `DmLogLevel(value int)`.

**Comment by runcom** [comment]
cc @nalind for c/storage (thanks for the ping!)

File: pkg/devicemapper/devmapper_wrapper.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper_wrapper.go]: I get following warning during compilation.

Building: bundles/17.06.0-dev/dynbinary-daemon/dockerd-17.06.0-dev
# github.com/docker/docker/pkg/devicemapper
.gopath/src/github.com/docker/docker/pkg/devicemapper/devmapper_wrapper.go: In function ‘log_cb’:
.gopath/src/github.com/docker/docker/pkg/devicemapper/devmapper_wrapper.go:20:2: warning: implicit declaration of function ‘vasprintf’; did you mean ‘vsprintf’? [-Wimplicit-function-declaration]
  vasprintf(&buffer, f, ap);
  ^~~~~~~~~
  vsprintf

Comment (rhvgoyal) [inline pkg/devicemapper/devmapper_wrapper.go]: Adding `#define _GNU_SOURCE` seems to make it go away.

Comment (cyphar) [inline pkg/devicemapper/devmapper_wrapper.go]: Yup. For some reason I didn't get the same warning. Hmm.

Comment (cyphar) [inline pkg/devicemapper/devmapper_wrapper.go]: Fixed.

**Comment by rhvgoyal** [comment]
I have disabled deferred removal and deferred deletion. And I launch a container and do "unshare -m" in a second window to make sure container mount point leaks. And then exit the container. I expected that container exit will try to remove device in a busy loop but exit seemed to happen immediately. I don't think its because of your patches but something has changed. Previously it did not behave like this.

BTW, how do I test this. How do I force an error condition so that I can see libdm messages in journal.

File: pkg/devicemapper/devmapper.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper.go]: How do we know that there are no external users of devicemapper package which are not making use of LogInitVerbose.

I think that might have been the thought in my mind and left LogInitVerbose() behind.

File: pkg/devicemapper/devmapper.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper.go]: I am again concerned about somebody using devicemapper package and using this interface and they will be broken now. 

I think rancher folks were using pkg/devicemapper.

This change does not seem fundamental to what you are trying to do. So why not first focus on the core change instead of removing existing interfaces of pkg/devicemapper/

File: daemon/graphdriver/devmapper/deviceset.go
Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/deviceset.go]: Somewhere in the comment it will be useful to put what values of libdm_log_level users can pass and what these map to.

File: pkg/devicemapper/devmapper.go
Comment (cyphar) [inline pkg/devicemapper/devmapper.go]: `LogInitVerbose` is meaningless if you are using callbacks. It only has meaning for the default logger, which we don't use because we need callbacks to get errors.

File: daemon/graphdriver/devmapper/deviceset.go
Comment (cyphar) [inline daemon/graphdriver/devmapper/deviceset.go]: I am planning on putting that in the CLI docs, but they map precisely to the `libdm` log levels.

File: pkg/devicemapper/devmapper.go
Comment (cyphar) [inline pkg/devicemapper/devmapper.go]: The semantics of this function never made sense because you **need** to have callbacks registered for this package to function. Which meant that a user of this package has to explicitly call this function so that the package will work as intended. That doesn't seem like a sane interface to me.

If you like, I can re-add `LogInit` but change it's semantics so that it acts as an additional logger (or a replacement for the `logrus` one we have now). But this will also "break" users in a more drastic way (rather than not compiling, code that worked previously will have slightly different semantics).

As for breaking people, surely everyone uses `vendor` these days so I'm not really sure what you mean by this. Yes, when they update they will have to make **very minor** changes but the alternative is to keep an old interface that will lead people to think that the semantics are unchanged (or that the semantics are different to what they actually are).

**Comment by cyphar** [comment]
> BTW, how do I test this. How do I force an error condition so that I can see libdm messages in journal.

If you enable debug logging you'll see all of the debug messages, just use `--storage-opt dm.libdm_log_level=7 --debug` and you'll see plenty of `logrus` debug messages fly by when you start and stop containers. If you forcefully kill the daemon you also might see some `libdm` errors as well.

**Comment by cyphar** [comment]
> And I launch a container and do "unshare -m" in a second window to make sure container mount point leaks. And then exit the container. I expected that container exit will try to remove device in a busy loop but exit seemed to happen immediately. I don't think its because of your patches but something has changed. Previously it did not behave like this.

I don't think that my patches could've caused that. Did you make sure you remounted `/` with `mount --make-rprivate /`?

File: pkg/devicemapper/devmapper.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper.go]: When you say `We`, you mean docker. I am concerned about other tools (outside docker) making use of devicemapper package. They can make use of it, right?

Comment (rhvgoyal) [inline pkg/devicemapper/devmapper.go]: I am not saying somebody is making use of it. Just that it is an interface exported by package and if we remove an existing interface (whether it makes much sense or not), there is always potential to break an existing user out there somewhere.

Comment (cyphar) [inline pkg/devicemapper/devmapper.go]: > When you say we, you mean docker.

No, I'm not talking about Docker, I'm talking about any potential user of this `pkg`.

`LogInitVerbose` *literally does not mean anything outside of the default `libdm` logger*, and because we (or any user of this package) **need** to set our own callbacks in order to get `errBusy` (which we use internally for a bunch of things) and so on, exposing this API is confusing and pointless. If you register your own logging callback, you get all messages without any filter.

If believe I'm mistaken, you can look at the source of ``:

```c
void dm_log_init_verbose(int level)
{
	_verbose = level;
}
```

And `_verbose` is only used in the default logger:

```
% git grep '\b_verbose\b' libdm/
libdm/libdm-common.c:static int _verbose = 0;
libdm/libdm-common.c:   if (level <= _LOG_WARN || _verbose) {
libdm/libdm-common.c:   _verbose = level;
```

```c
/*
 * Library users can provide their own logging
 * function.
 */

__attribute__((format(printf, 5, 0)))
static void _default_log_line(int level,
	    const char *file __attribute__((unused)),
	    int line __attribute__((unused)), int dm_errno_or_class,
	    const char *f, va_list ap)
{
	static int _abort_on_internal_errors = -1;
	FILE *out = log_stderr(level) ? stderr : stdout;

	level = log_level(level);

	if (level <= _LOG_WARN || _verbose) {
		if (level < _LOG_WARN)
			out = stderr;
		vfprintf(out, f, ap);
		fputc('\n', out);
	}

	if (_abort_on_internal_errors < 0)
		/* Set when env DM_ABORT_ON_INTERNAL_ERRORS is not "0" */
		_abort_on_internal_errors =
			strcmp(getenv("DM_ABORT_ON_INTERNAL_ERRORS") ? : "0", "0");

	if (_abort_on_internal_errors &&
	    !strncmp(f, INTERNAL_ERROR, sizeof(INTERNAL_ERROR) - 1))
		abort();
}
```

And when you register your own logger, it replaces the default one. I checked all of this before removing this API, because I was also confused why it was present in this library. If anyone is using it, it's acting as a noop.

File: pkg/devicemapper/devmapper.go
Comment (cyphar) [inline pkg/devicemapper/devmapper.go]: Just to be clear, I do understand wanted to provide logging hooks, so maybe we should re-add it. I'm just trying to point out that the old API was not correct. Sorry for the strong reply.

File: pkg/devicemapper/devmapper.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper.go]: Ok, given other core interfaces like `RemoveDevice` and `DeleteDevice` rely on `dmSawBusy` and it will be set only if user of package calls `LogInit()` (Either with or without a logger). Otherwise package will always return error even in case of busy device and send messages from libdm to stdout.

IOW, for devicemapper pkg to work properly, this package always need to register a callback function with libdm. And right now unfortunately it is initiated from the user of package.

So I agree that cleaning this up makes sense. While there might be a very strange user out there but given how twisted this internface is, chances of this happening are small. If it does happen, we will try to advise them to move to new interface.

So, I am fine with removing `LogInitVerbose`. Please do capture some of this information in changelog for future reference.

File: pkg/devicemapper/devmapper.go
Comment (rhvgoyal) [inline pkg/devicemapper/devmapper.go]: I like the idea of retaining `LogInit()` but changing the semantics so that caller can register its own logger and get all the messages. IOW, it replaced logrus. 

So if existing users were using it without any logger `LogInit()`, they will not see any change except that logs will start going to journal. Those who were registering their own logger, they will also not see any change as they will continue to get all the logs and then they can decide what to do with those logs.

IOW, default logger of pkg/devicemapper will be `logrus` until and unless overwridden by `LogInit`. 

Does this sound reasonable to you?

File: pkg/devicemapper/devmapper.go
Comment (cyphar) [inline pkg/devicemapper/devmapper.go]: Will do, sorry for the semi-long rant. I cannot describe in words how silly I find it that `libdm` requires you to register a logging callback in order to understand what an error code of `1` actually means.

File: pkg/devicemapper/devmapper.go
Comment (cyphar) [inline pkg/devicemapper/devmapper.go]: That sounds fine to me, I'll update it accordingly.

**Comment by cyphar** [comment]
Alright, I've fixed up your concerns @rhvgoyal. WDYT?

**Comment by rhvgoyal** [comment]
LGTM

@cyphar thanks for implementing this knob. This is really required to be able to debug some docker issues without recompiling.

Rest of the cleanup is also nice. Makes reading this code easier.

**Comment by cyphar** [comment]
I'm using this patchset to debug some production issues right now. :wink:

**Comment by cyphar** [comment]
Tests pass, and everything seems to work.

/ping @cpuguy83 @vdemeester

**Comment by coolljt0725** [comment]
ping @cpuguy83

**Review by cpuguy83 — APPROVED** [review]
LGTM

Tested, looks good.

**Comment by cyphar** [comment]
This has been approved by two maintainers, is there anyone else we need to ping @coolljt0725?

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
thanks @cyphar !

**Pull request merged**
